### PR TITLE
Implement `eth_call` operation

### PR DESCRIPTION
**Pull request by MathieuDutSik**
## Motivation

Accessing Ethereum smart contract with `eth_call` allows to run a contract without changing the state of the blockchain. This is useful for accessing functions with no side effects such as `balanceOf`.

## Proposal

The following was done:
* The smart contract was split into two for clarity.
* The function `non_executive_call` is added to the `client`.
* The contracts are implemented as objects and called as such from the tests.
* The functions is added to the Oracle code and so are serializable/deserializable.
* The function calls follow the ethers code with a relatively simple API.

## Test Plan

The tests have been restructured and strenghtened.

## Release Plan

Not relevant yet.

## Links

- [reviewer checklist](https://github.com/linera-io/linera-protocol/blob/main/CONTRIBUTING.md#reviewer-checklist)

### CODE REVIEW

**Review by afck — APPROVED** [review]
Looks good, but we might have to modify quite a lot of this later to make it work with the `fetch_json` approach.

File: linera-ethereum/contracts/event_numerics.sol [hunk 1]
```diff
@@ -0,0 +1,32 @@
+/// SPDX-License-Identifier: UNLICENSED
```
[the remark this answers is not in the document]
Comment (MathieuDutSik) [reply linera-ethereum/contracts/event_numerics.sol:1]: Yes, correct, thank you.

File: linera-ethereum/src/client.rs [hunk 2]
```diff
@@ -77,6 +78,7 @@ impl EthereumEndpoint {
     ) -> Result<Vec<EthereumEvent>, EthereumServiceError> {
         let contract_address = contract_address.parse::<Address>()?;
         let event_name = event_name_from_expanded(event_name_expanded);
+        println!("event_name={}", event_name);
```
[the remark this answers is not in the document]
Comment (MathieuDutSik) [reply linera-ethereum/src/client.rs:81]: No, of course. Sorry for that.

File: linera-ethereum/src/client.rs [hunk 3]
```diff
@@ -88,4 +90,24 @@ impl EthereumEndpoint {
             .collect::<Result<_, _>>()?;
         Ok(events)
     }
+
+    /// The operation done with `eth_call` on the Ethereum returns
```
[the remark this answers is not in the document]
Comment (MathieuDutSik) [reply linera-ethereum/src/client.rs:94]: Ok.

File: linera-ethereum/tests/ethereum_test.rs [hunk 4]
```diff
@@ -59,13 +57,42 @@ async fn test_contract() -> anyhow::Result<()> {
         ],
         block_number: 1,
     };
+    assert_eq!(events.len(), 1);
     for event in events {
         assert_eq!(event, target_event);
     }
```
[the remark this answers is not in the document]
Comment (MathieuDutSik) [reply linera-ethereum/tests/ethereum_test.rs:63]: Ok, thank you very much.

File: linera-ethereum/src/client.rs [hunk 5]
```diff
@@ -88,4 +90,24 @@ impl EthereumEndpoint {
...
+    /// a result but are not executed. This can be useful for example
+    /// for executing function that are const and allow to inspect
+    /// the contract without modifying it.
+    pub async fn non_executive_call(
+        &self,
+        contract_address: &str,
+        data: Bytes,
+        from: &str,
```
[the remark this answers is not in the document]
Comment (MathieuDutSik) [reply linera-ethereum/src/client.rs:101]: Yes, there are further cleaning issues to address. But I propose to do that later.

**Review by afck — APPROVED** [review]

**Pull request merged**